### PR TITLE
Replace strp values with references to the strings in the debug_str section

### PR DESCRIPTION
**Pull request by rastogishubham**
To improve deduplication in the compile units but to not increase the size of the distinct data, add all strings that do dedupe across translation units as cas references to the compile unit rather than move all strps into the distinct data.

Only DW_AT_name and DW_AT_linkage_name are the strps that do not dedupe and are moved into the distinct data.

### CODE REVIEW

**Comment by rastogishubham** [comment]
@swift-ci please test

**Comment by rastogishubham** [comment]
@swift-ci please test

File: llvm/include/llvm/MC/CAS/MCCASObjectV1.h [hunk 1]
```diff
@@ -145,6 +145,13 @@ class MCObjectProxy : public cas::ObjectProxy {
 
   MCObjectProxy() = delete;
 
+  static Error encodeReferences(ArrayRef<cas::ObjectRef> Refs,
+                                SmallVectorImpl<char> &Data,
+                                SmallVectorImpl<cas::ObjectRef> &IDs);
+
+  static Expected<SmallVector<cas::ObjectRef>>
+  decodeReferences(const MCObjectProxy &Node, StringRef &Remaining);
```
Comment (felipepiovezan) [inline llvm/include/llvm/MC/CAS/MCCASObjectV1.h:153]: This kind of refactor should go into its own commit, with some explanation on the commit message as to why we're doing it

File: llvm/include/llvm/MC/CAS/MCCASObjectV1.h [hunk 2]
```diff
@@ -543,6 +588,13 @@ class MCCASBuilder {
   splitAbbrevSection(ArrayRef<size_t> AbbrevOffsets,
                      ArrayRef<char> FullAbbrevData);
 
+  struct DebugStringSectionContents {
+    SmallVector<DebugStrRef, 0> DebugStringRefs;
+    static DenseMap<unsigned, cas::ObjectRef> MapOfStringRefs;
```
Comment (felipepiovezan) [inline llvm/include/llvm/MC/CAS/MCCASObjectV1.h]: Why is this static? I think we should find a proper ownership location for this, instead of relying on global state.
If the compiler is used in a library mode, this would create difficult to debug problems

File: llvm/include/llvm/MC/CAS/MCCASObjectV1.h [hunk 3]
```diff
@@ -632,10 +684,12 @@ class DebugInfoSectionRef : public SpecificRef<DebugInfoSectionRef> {
...
   }
-  Expected<uint64_t> materialize(MCCASReader &Reader,
-                                 ArrayRef<char> AbbrevSectionContents,
-                                 ArrayRef<uint32_t> SecOffsetVals,
-                                 raw_ostream *Stream = nullptr) const;
+
+  Expected<uint64_t>
+  materialize(MCCASReader &Reader, ArrayRef<char> AbbrevSectionContents,
+              ArrayRef<uint32_t> SecOffsetVals,
+              DenseMap<cas::ObjectRef, unsigned> &MapOfStringOffsets,
```
Comment (felipepiovezan) [inline llvm/include/llvm/MC/CAS/MCCASObjectV1.h]: Do we mutate `MapStringOffsets` during materialization?

File: llvm/lib/MC/MCCASObjectV1.cpp [hunk 4]
```diff
@@ -554,9 +584,56 @@ LoadedDebugLineSection::load(DebugLineSectionRef Section) {
...
+  Error addNode(Expected<MCObjectProxy> Proxy, unsigned &StringOffset) {
+    if (!Proxy)
+      return Proxy.takeError();
+    if (auto StrRef = DebugStrRef::Cast(*Proxy)) {
+      MapOfStringOffsets.try_emplace(StrRef->getRef(), StringOffset);
+      StringOffset += StrRef->getData().size() + 1;
+    } else
+      return createStringError(inconvertibleErrorCode(),
+                               "Invalid node for Debug Info section");
+    return Error::success();
```
Comment (felipepiovezan) [inline llvm/lib/MC/MCCASObjectV1.cpp]: we can simplify control flow: let's move the return into the `if` and eliminate the `else`

File: llvm/lib/MC/MCCASObjectV1.cpp [hunk 5]
```diff
@@ -554,9 +584,56 @@ LoadedDebugLineSection::load(DebugLineSectionRef Section) {
...
 }
 
+struct LoadedDebugStringSection {
+  DenseMap<cas::ObjectRef, unsigned> MapOfStringOffsets;
+
+  static Expected<LoadedDebugStringSection> load(DebugStringSectionRef Section);
+
+private:
+  /// Extracts the information contained in \p Proxy, which must be either an
+  /// StrRef or a PaddingNode.
```
Comment (felipepiovezan) [inline llvm/lib/MC/MCCASObjectV1.cpp]: The comments mention `PaddingNode`, but we don't see that in the code?

File: llvm/lib/MC/MCCASObjectV1.cpp [hunk 6]
```diff
@@ -554,9 +584,56 @@ LoadedDebugLineSection::load(DebugLineSectionRef Section) {
...
+  Error addNode(Expected<MCObjectProxy> Proxy, unsigned &StringOffset) {
+    if (!Proxy)
+      return Proxy.takeError();
+    if (auto StrRef = DebugStrRef::Cast(*Proxy)) {
+      MapOfStringOffsets.try_emplace(StrRef->getRef(), StringOffset);
+      StringOffset += StrRef->getData().size() + 1;
```
Comment (felipepiovezan) [inline llvm/lib/MC/MCCASObjectV1.cpp:597]: Why the +1 here? Aren't DWARF strings null terminated?

File: llvm/lib/MC/MCCASObjectV1.cpp [hunk 7]
```diff
@@ -554,9 +584,56 @@ LoadedDebugLineSection::load(DebugLineSectionRef Section) {
...
+    return MCObjectProxy::get(Schema, CAS.getProxy(Ref));
+  };
+
+  LoadedDebugStringSection LoadedSection;
+  // TODO: Add support for 64 bit dwarf format
+  unsigned StringOffset = 0;
+  for (auto Ref : *Refs) {
+    if (auto E = LoadedSection.addNode(loadRef(Ref), StringOffset))
+      return std::move(E);
+  }
```
Comment (felipepiovezan) [inline llvm/lib/MC/MCCASObjectV1.cpp]: no need for braces here

File: llvm/include/llvm/MC/CAS/MCCASObjectV1.def [hunk 8]
```diff
@@ -18,7 +18,6 @@ CASV1_SIMPLE_DATA_REF(DataInCodeRef, mc:data_in_code)
 CASV1_SIMPLE_DATA_REF(CStringRef, mc:cstring)
 CASV1_SIMPLE_DATA_REF(MergedFragmentRef, mc:merged_fragment)
 CASV1_SIMPLE_DATA_REF(DebugStrRef, mc:debug_string)
-CASV1_SIMPLE_DATA_REF(DebugInfoCURef, mc:debug_info_cu)
```
Comment (felipepiovezan) [inline llvm/include/llvm/MC/CAS/MCCASObjectV1.def:21]: This would also be easier to review in a separate commit :)

File: llvm/lib/MC/MCCASObjectV1.cpp [hunk 9]
```diff
@@ -775,7 +878,14 @@ struct LoadedDebugInfoSection {
...
         return createStringError(inconvertibleErrorCode(),
                                  "Invalid CURef after PaddingRef");
-      CUData.push_back(CURef->getData());
+
+      StringRef Remaining = CURef->getData();
+      auto DebugStrRefs = MCObjectProxy::decodeReferences(*Proxy, Remaining);
+      if (!DebugStrRefs)
+        return DebugStrRefs.takeError();
+      DebugStringRefs.resize(DebugStringRefs.size() + 1);
+      DebugStringRefs.back().append(*DebugStrRefs);
```
Comment (felipepiovezan) [inline llvm/lib/MC/MCCASObjectV1.cpp]: Subjective, but I think this is more cleanly expressed as `DebugStringRefs.push_back(std::move(*DebugStrRefs));`

File: llvm/lib/MC/MCCASObjectV1.cpp [hunk 10]
```diff
@@ -775,7 +878,14 @@ struct LoadedDebugInfoSection {
       if (Padding.has_value())
         return createStringError(inconvertibleErrorCode(),
                                  "Invalid CURef after PaddingRef");
-      CUData.push_back(CURef->getData());
+
+      StringRef Remaining = CURef->getData();
+      auto DebugStrRefs = MCObjectProxy::decodeReferences(*Proxy, Remaining);
```
Comment (felipepiovezan) [inline llvm/lib/MC/MCCASObjectV1.cpp:879]: We can bypass the complexity of having to encode/decode references: we never refer to the same strp twice by design, so the encoding/decoding will never do anything, as it is designed to compress multiple edges to the same block.

File: llvm/lib/MC/MCCASObjectV1.cpp [hunk 11]
```diff
@@ -1158,6 +1288,39 @@ Expected<uint64_t> DebugLineSectionRef::materialize(MCCASReader &Reader,
...
+Expected<uint64_t>
+DebugStringSectionRef::materialize(MCCASReader &Reader,
+                                   raw_ostream *Stream) const {
+  // Start a new section for relocations.
+  Reader.Relocations.emplace_back();
+  SmallVector<char, 0> SectionContents;
+  raw_svector_ostream SectionStream(SectionContents);
+
+  unsigned Size = 0;
+  StringRef Remaining = getData();
```
Comment (felipepiovezan) [inline llvm/lib/MC/MCCASObjectV1.cpp:1295]: should we assert `Remaining.empty()` at the end of this function?

File: llvm/lib/MC/MCCASObjectV1.cpp (hunk 10 above)
Comment (felipepiovezan) [reply llvm/lib/MC/MCCASObjectV1.cpp:879]: I see now that we _are_ in fact storing repeated edges

File: llvm/lib/MC/MCCASObjectV1.cpp [hunk 12]
```diff
@@ -1898,18 +2064,29 @@ static void partitionCUDie(
...
       append_range(PartitionedData.DistinctData,
                    DebugInfoData.slice(AttrValue.Offset, AttrValue.ByteSize));
-    else if (AttrValue.Attr != llvm::dwarf::DW_AT_stmt_list)
+    } else if (AttrValue.Value.getForm() == llvm::dwarf::DW_FORM_strp) {
+      // TODO: Add support for 64-bit DWARF
+      uint32_t StringOffset;
+      memcpy(&StringOffset, &DebugInfoData[AttrValue.Offset], 4);
+      auto StringRef = Builder.getObjectRefFromStringMap(StringOffset);
+      if (!StringRef)
+        llvm_unreachable("Strp value not found in debug string section!");
```
Comment (felipepiovezan) [inline llvm/lib/MC/MCCASObjectV1.cpp]:
```
if (cond)
   unreachable
```

->

```
assert(!cond)
```

**Comment by rastogishubham** [comment]
@swift-ci please test

File: llvm/lib/MC/MCCASObjectV1.cpp [hunk 13]
```diff
@@ -1920,9 +2097,11 @@ static void partitionCUDie(
 }
 
 Expected<InMemoryCASDWARFObject::PartitionedDebugInfoSection>
-InMemoryCASDWARFObject::partitionCUData(ArrayRef<char> DebugInfoData,
+InMemoryCASDWARFObject::partitionCUData(MCCASBuilder &Builder,
+                                        ArrayRef<char> DebugInfoData,
                                         uint64_t AbbrevOffset,
-                                        uint64_t CUOffset, DWARFContext *Ctx) {
+                                        uint64_t CUOffset, DWARFContext *Ctx,
+                                        ArrayRef<DebugStrRef> DebugStringRefs) {
```
Comment (felipepiovezan) [inline llvm/lib/MC/MCCASObjectV1.cpp]: Is `DebugStringRefs` used anywhere in this function?

**Review by felipepiovezan — COMMENTED** [review]
One high level question: is this really bringing the total size of the CAS down?

Let's say we have 100 strps that deduplicate in some CU. If we do this, we're going from 400 bytes to 800 bytes worth of strps. So, if we deduplicate the CU at least once, we break even and we are profitable afterwards.

An alternative would be to LEB encode the _index_ into the list of references in the string table object. This index would take one or two bytes in most cases. However, the index would need to be stored in the distinct data block, and would not deduplicate.

Could you post some numbers of the total CAS size before/after this patch?

**Comment by rastogishubham** [comment]
> One high level question: is this really bringing the total size of the CAS down?
> 
> Let's say we have 100 strps that deduplicate in some CU. If we do this, we're going from 400 bytes to 800 bytes worth of strps. So, if we deduplicate the CU at least once, we break even and we are profitable afterwards.
> 
> An alternative would be to LEB encode the _index_ into the list of references in the string table object. This index would take one or two bytes in most cases. However, the index would need to be stored in the distinct data block, and would not deduplicate.
> 
> Could you post some numbers of the total CAS size before/after this patch?

Full Debug Info without strp as ref patch:
builtin:node                   38   0.00%     18,064   0.00%         73   0.00%        506   0.00%      2,306   0.00%
builtin:tree                5,782   0.07%      6,513   0.00%     29,784   0.01%  1,153,208   0.01%  1,576,504   0.01%
mc:align                       30   0.00%     24,964   0.01%          0   0.00%        210   0.00%      1,170   0.00%
mc:assembler               17,991   0.23%     23,271   0.01%     89,955   0.02%    485,757   0.00%  1,781,109   0.01%
mc:atom                 3,426,763  43.88%  9,649,905   2.33%  4,330,008   1.04% 433,552,133   2.55% 577,848,613   2.81%
mc:cstring                935,924  11.98%  4,592,292   1.11%          0   0.00% 115,695,491   0.68% 145,645,059   0.71%
mc:data                    91,884   1.18%    339,457   0.08%          0   0.00% 2,329,924,347  13.71% 2,332,864,635  11.34%
mc:data_in_code             8,394   0.11%     17,991   0.00%          0   0.00% 212,681,450   1.25% 212,950,058   1.04%
mc:debug_abbrev            23,668   0.30%  1,084,740   0.26%          0   0.00% 17,584,308   0.10% 18,341,684   0.09%
mc:debug_abbrev_offsets      3,784   0.05%     17,638   0.00%          0   0.00% 14,186,705   0.08% 14,307,793   0.07%
mc:debug_abbrev_section      5,244   0.07%     17,990   0.00%  1,089,984   0.26%  9,809,315   0.06% 18,696,995   0.09%
mc:debug_info_cu          131,797   1.69% 21,569,087   5.20%          0   0.00% 6,236,581,942  36.69% 6,240,799,446  30.35%
mc:debug_info_distinct_data     17,580   0.23%     17,638   0.00%          0   0.00% 6,103,881,313  35.90% 6,104,443,873  29.68%
mc:debug_info_section      17,638   0.23%     17,990   0.00% 21,622,001   5.22% 668,760,771   3.93% 842,301,195   4.10%
mc:debug_line             211,222   2.70%  8,346,857   2.01%          0   0.00% 100,094,257   0.59% 106,853,361   0.52%
mc:debug_line_section      11,118   0.14%     17,990   0.00%  8,357,975   2.02% 264,268,741   1.55% 331,488,317   1.61%
mc:debug_string         2,318,625  29.69% 364,368,219  87.90%          0   0.00% 275,736,684   1.62% 349,932,684   1.70%
mc:fill                       205   0.00%     11,834   0.00%          0   0.00%        931   0.00%      7,491   0.00%
mc:group                   17,991   0.23%     17,991   0.00%    298,891   0.07%     35,987   0.00%  3,002,827   0.01%
mc:header                  15,989   0.20%     17,991   0.00%          0   0.00% 26,052,909   0.15% 26,564,557   0.13%
mc:leb                          6   0.00%         17   0.00%          0   0.00%         12   0.00%        204   0.00%
mc:merged_fragment        454,273   5.82%  3,953,736   0.95%          0   0.00% 179,080,988   1.05% 193,617,724   0.94%
mc:padding                     16   0.00%    113,036   0.03%          0   0.00%         32   0.00%        544   0.00%
mc:section                 91,020   1.17%    244,921   0.06% 374,097,160  90.25% 10,536,987   0.06% 3,006,226,907  14.62%
mc:symbol_table             3,263   0.04%     17,991   0.00%  4,592,292   1.11%      9,378   0.00% 36,852,130   0.18%
TOTAL                   7,810,245 100.00% 414,508,123 100.00% 414,508,123 100.00% 17,000,114,362 100.00% 20,566,107,186 100.00%
Full Debug info with strp as refs patch:
builtin:node                   39   0.00%     18,066   0.00%         75   0.00%        529   0.00%      2,377   0.00%
builtin:tree                5,782   0.07%      6,513   0.00%     29,784   0.01%  1,153,208   0.01%  1,576,504   0.01%
mc:align                       30   0.00%     25,143   0.01%          0   0.00%        210   0.00%      1,170   0.00%
mc:assembler               17,991   0.23%     23,271   0.01%     89,955   0.02%    485,757   0.00%  1,781,109   0.01%
mc:atom                 3,424,631  43.86%  9,649,865   2.33%  4,330,657   1.04% 434,238,003   2.58% 578,471,451   2.84%
mc:cstring                935,924  11.99%  4,592,292   1.11%          0   0.00% 115,695,491   0.69% 145,645,059   0.71%
mc:data                    91,948   1.18%    346,014   0.08%          0   0.00% 2,330,200,197  13.84% 2,333,142,533  11.44%
mc:data_in_code             8,405   0.11%     17,991   0.00%          0   0.00% 214,334,565   1.27% 214,603,525   1.05%
mc:debug_abbrev            23,668   0.30%  1,084,740   0.26%          0   0.00% 17,584,308   0.10% 18,341,684   0.09%
mc:debug_abbrev_offsets      3,784   0.05%     17,631   0.00%          0   0.00% 14,186,705   0.08% 14,307,793   0.07%
mc:debug_abbrev_section      5,244   0.07%     17,990   0.00%  1,089,984   0.26%  9,809,315   0.06% 18,696,995   0.09%
mc:debug_info_cu          131,596   1.69% 21,555,144   5.20%    298,867   0.07% 6,236,758,953  37.05% 6,243,360,961  30.61%
mc:debug_info_distinct_data     17,580   0.23%     17,631   0.00%          0   0.00% 5,931,364,337  35.24% 5,931,926,897  29.08%
mc:debug_info_section      17,631   0.23%     17,990   0.00% 21,608,037   5.21% 668,378,842   3.97% 841,807,330   4.13%
mc:debug_line             210,966   2.70%  8,350,298   2.01%          0   0.00% 100,646,218   0.60% 107,397,130   0.53%
mc:debug_line_section      11,118   0.14%     17,990   0.00%  8,361,416   2.02% 264,305,061   1.57% 331,552,165   1.63%
mc:debug_string         2,318,625  29.70% 364,667,086  87.91%          0   0.00% 276,296,958   1.64% 350,492,958   1.72%
mc:debug_string_section     11,984   0.15%     17,990   0.00% 364,368,219  87.84%     47,936   0.00% 2,915,377,176  14.29%
mc:fill                       205   0.00%     11,843   0.00%          0   0.00%        931   0.00%      7,491   0.00%
mc:group                   17,991   0.23%     17,991   0.00%    298,891   0.07%     35,987   0.00%  3,002,827   0.01%
mc:header                  15,989   0.20%     17,991   0.00%          0   0.00% 26,052,909   0.15% 26,564,557   0.13%
mc:leb                          6   0.00%         17   0.00%          0   0.00%         12   0.00%        204   0.00%
mc:merged_fragment        454,180   5.82%  3,947,640   0.95%          0   0.00% 179,163,154   1.06% 193,696,914   0.95%
mc:padding                     16   0.00%    113,075   0.03%          0   0.00%         32   0.00%        544   0.00%
mc:section                 79,082   1.01%    226,931   0.05%  9,728,947   2.35% 10,489,229   0.06% 90,851,429   0.45%
mc:symbol_table             3,263   0.04%     17,991   0.00%  4,592,292   1.11%      9,378   0.00% 36,852,130   0.18%
TOTAL                   7,807,678 100.00% 414,797,124 100.00% 414,797,124 100.00% 16,831,238,225 100.00% 20,399,460,913 100.00%
[10:57](https://a1391192.slack.com/archives/D03P2TBB8CB/p1668711451875009)
-gline-tables-only without strps as refs:
builtin:node                   38   0.00%     11,101   0.03%         73   0.00%        506   0.00%      2,306   0.00%
builtin:tree                4,698   0.09%      5,754   0.02%     28,104   0.08%  1,094,248   0.07%  1,469,416   0.07%
mc:align                       30   0.00%     24,963   0.07%          0   0.00%        210   0.00%      1,170   0.00%
mc:assembler               11,028   0.20%     22,350   0.07%     55,140   0.16%    297,756   0.02%  1,091,772   0.05%
mc:atom                 3,374,203  62.59%  9,596,355  28.65%  3,862,386  11.53% 299,920,919  18.38% 438,794,503  21.18%
mc:cstring                935,910  17.36%  4,590,803  13.70%          0   0.00% 115,693,176   7.09% 145,642,296   7.03%
mc:data                    49,451   0.92%    296,932   0.89%          0   0.00% 365,598,621  22.41% 367,181,053  17.72%
mc:data_in_code             3,703   0.07%     11,028   0.03%          0   0.00% 90,102,311   5.52% 90,220,807   4.35%
mc:debug_abbrev                18   0.00%      3,005   0.01%          0   0.00%      1,237   0.00%      1,813   0.00%
mc:debug_abbrev_offsets      1,809   0.03%      3,388   0.01%          0   0.00%  8,928,411   0.55%  8,986,299   0.43%
mc:debug_abbrev_section      2,031   0.04%     11,017   0.03%      5,036   0.02%  3,481,955   0.21%  3,587,235   0.17%
mc:debug_info_cu            5,297   0.10%  4,727,000  14.11%          0   0.00%    227,293   0.01%    396,797   0.02%
mc:debug_info_distinct_data      3,194   0.06%      3,388   0.01%          0   0.00% 67,811,554   4.16% 67,913,762   3.28%
mc:debug_info_section       3,388   0.06%     11,017   0.03%  4,737,164  14.14% 128,274,512   7.86% 166,280,240   8.03%
mc:debug_line             202,102   3.75%  8,000,472  23.88%          0   0.00% 60,917,788   3.73% 67,385,052   3.25%
mc:debug_line_section      10,708   0.20%     11,017   0.03%  8,011,180  23.92% 256,772,682  15.74% 321,204,778  15.50%
mc:debug_string           278,170   5.16%  2,412,508   7.20%          0   0.00% 35,318,958   2.16% 44,220,398   2.13%
mc:fill                       205   0.00%     11,834   0.04%          0   0.00%        931   0.00%      7,491   0.00%
mc:group                   11,026   0.20%     11,028   0.03%    171,302   0.51%     22,111   0.00%  1,745,359   0.08%
mc:header                   5,014   0.09%     11,028   0.03%          0   0.00%  7,636,942   0.47%  7,797,390   0.38%
mc:leb                          6   0.00%         17   0.00%          0   0.00%         12   0.00%        204   0.00%
mc:merged_fragment        454,222   8.43%  3,528,640  10.53%          0   0.00% 179,030,905  10.97% 193,566,009   9.34%
mc:padding                     16   0.00%     43,719   0.13%          0   0.00%         32   0.00%        544   0.00%
mc:section                 31,592   0.59%    138,251   0.41% 12,036,455  35.93% 10,292,211   0.63% 107,594,795   5.19%
mc:symbol_table             3,263   0.06%     11,028   0.03%  4,590,803  13.70%      9,385   0.00% 36,840,225   1.78%
TOTAL                   5,391,122 100.00% 33,497,643 100.00% 33,497,643 100.00% 1,631,434,666 100.00% 2,071,931,714 100.00%
-gline-tables-only with strps as refs:
builtin:node                   39   0.00%     11,104   0.03%         75   0.00%        529   0.00%      2,377   0.00%
builtin:tree                4,698   0.09%      5,754   0.02%     28,104   0.08%  1,094,248   0.07%  1,469,416   0.07%
mc:align                       30   0.00%     25,143   0.08%          0   0.00%        210   0.00%      1,170   0.00%
mc:assembler               11,029   0.20%     22,350   0.07%     55,145   0.16%    297,783   0.02%  1,091,871   0.05%
mc:atom                 3,372,140  62.57%  9,597,756  28.64%  3,860,492  11.52% 299,696,659  18.74% 438,489,075  21.50%
mc:cstring                935,916  17.37%  4,592,645  13.70%          0   0.00% 115,693,623   7.23% 145,642,935   7.14%
mc:data                    49,517   0.92%    303,518   0.91%          0   0.00% 365,897,935  22.88% 367,482,479  18.02%
mc:data_in_code             3,701   0.07%     11,029   0.03%          0   0.00% 90,728,941   5.67% 90,847,373   4.45%
mc:debug_abbrev                18   0.00%      3,005   0.01%          0   0.00%      1,237   0.00%      1,813   0.00%
mc:debug_abbrev_offsets      1,809   0.03%      3,388   0.01%          0   0.00%  8,928,411   0.56%  8,986,299   0.44%
mc:debug_abbrev_section      2,031   0.04%     11,018   0.03%      5,036   0.02%  3,481,955   0.22%  3,587,235   0.18%
mc:debug_info_cu            5,339   0.10%  4,723,209  14.09%     10,696   0.03%    234,354   0.01%    490,770   0.02%
mc:debug_info_distinct_data      3,195   0.06%      3,388   0.01%          0   0.00% 33,917,795   2.12% 34,020,035   1.67%
mc:debug_info_section       3,388   0.06%     11,018   0.03%  4,733,373  14.12% 128,175,867   8.02% 166,151,267   8.15%
mc:debug_line             201,847   3.75%  8,005,357  23.89%          0   0.00% 61,412,875   3.84% 67,871,979   3.33%
mc:debug_line_section      10,709   0.20%     11,018   0.03%  8,016,066  23.92% 256,836,809  16.06% 321,308,025  15.75%
mc:debug_string           278,172   5.16%  2,423,965   7.23%          0   0.00% 35,614,538   2.23% 44,516,042   2.18%
mc:debug_string_section      4,001   0.07%     11,018   0.03%  2,413,269   7.20%     16,004   0.00% 19,450,188   0.95%
mc:fill                       205   0.00%     11,845   0.04%          0   0.00%        931   0.00%      7,491   0.00%
mc:group                   11,027   0.20%     11,029   0.03%    171,317   0.51%     22,113   0.00%  1,745,513   0.09%
mc:header                   5,013   0.09%     11,029   0.03%          0   0.00%  7,634,933   0.48%  7,795,349   0.38%
mc:leb                          6   0.00%         17   0.00%          0   0.00%         12   0.00%        204   0.00%
mc:merged_fragment        454,340   8.43%  3,519,969  10.50%          0   0.00% 179,123,730  11.20% 193,662,610   9.49%
mc:padding                     16   0.00%     43,720   0.13%          0   0.00%         32   0.00%        544   0.00%
mc:section                 27,592   0.51%    127,245   0.38%  9,625,348  28.72% 10,279,363   0.64% 88,165,091   4.32%
mc:symbol_table             3,264   0.06%     11,029   0.03%  4,592,645  13.70%      9,387   0.00% 36,854,995   1.81%
TOTAL                   5,389,042 100.00% 33,511,566 100.00% 33,511,566 100.00% 1,599,100,274 100.00% 2,039,642,146 100.00%

File: llvm/include/llvm/MC/CAS/MCCASObjectV1.h (hunk 3 above)
Comment (rastogishubham) [reply llvm/include/llvm/MC/CAS/MCCASObjectV1.h]: We don't, should I pass a pointer instead?

File: llvm/lib/MC/MCCASObjectV1.cpp (hunk 5 above)
Comment (rastogishubham) [reply llvm/lib/MC/MCCASObjectV1.cpp]: good catch, I probably copy pasted the comment and forgot to remove the paddingnode there, there are no paddingnodes in the debugstr section

File: llvm/lib/MC/MCCASObjectV1.cpp (hunk 6 above)
Comment (rastogishubham) [reply llvm/lib/MC/MCCASObjectV1.cpp:597]: They are, but we are splitting on '\0' and do not store the null terminator in the cas itself, that is why `StrRef->getData().size()` is the size of the string without the null terminator, which needs to be added

Comment (rastogishubham) [reply llvm/lib/MC/MCCASObjectV1.cpp:597]: Why take up extra bytes by storing the null terminator, when we can just add it back in materialization :)

File: llvm/lib/MC/MCCASObjectV1.cpp (hunk 9 above)
Comment (rastogishubham) [reply llvm/lib/MC/MCCASObjectV1.cpp:887]: I agree!

Comment (rastogishubham) [reply llvm/lib/MC/MCCASObjectV1.cpp]: Those two are not equivalent, `DebugStringRefs` is a `SmallVector<SmallVector<cas::ObjectRef>> DebugStringRefs;` I am appending the objectref to the last smallvector of objectrefs. Did you mean:

`DebugStringRefs.back().push_back(std::move(*DebugStrRefs));`?

File: llvm/lib/MC/MCCASObjectV1.cpp (hunk 11 above)
Comment (rastogishubham) [reply llvm/lib/MC/MCCASObjectV1.cpp:1295]: Yeah of course!

Comment (rastogishubham) [reply llvm/lib/MC/MCCASObjectV1.cpp:1295]: But we should do that separately, because we should be consistent and do it for all SectionRefs

File: llvm/lib/MC/MCCASObjectV1.cpp (hunk 13 above)
Comment (rastogishubham) [reply llvm/lib/MC/MCCASObjectV1.cpp]: You are right, it is not used, I changed the code to address that

File: llvm/include/llvm/MC/CAS/MCCASObjectV1.h (hunk 2 above)
Comment (rastogishubham) [reply llvm/include/llvm/MC/CAS/MCCASObjectV1.h]: It doesn't need to be static, I have changed it to no longer be static

**Comment by rastogishubham** [comment]
@swift-ci please test

File: llvm/include/llvm/MC/CAS/MCCASObjectV1.h (hunk 3 above)
Comment (felipepiovezan) [reply llvm/include/llvm/MC/CAS/MCCASObjectV1.h]: No, but we should add `const` in that case!

File: llvm/lib/MC/MCCASObjectV1.cpp (hunk 9 above)
Comment (felipepiovezan) [reply llvm/lib/MC/MCCASObjectV1.cpp]: > I am appending the objectref to the last smallvector of objectrefs

I don't see that. You are creating a new SmallVector:

> DebugStringRefs.resize(DebugStringRefs.size() + 1);

And then you append `*DebugStrRefs` to it. Append will take **_all_** the contents of of `*DebugStrRefs` and add them to the newly created vector.

So this:

> `DebugStringRefs.push_back(std::move(*DebugStrRefs))`

Is also doing the exact same thing. Does that make sense?

File: llvm/include/llvm/MC/CAS/MCCASObjectV1.h (hunk 1 above)
Comment (rastogishubham) [reply llvm/include/llvm/MC/CAS/MCCASObjectV1.h:153]: Done

File: llvm/include/llvm/MC/CAS/MCCASObjectV1.h (hunk 3 above)
Comment (rastogishubham) [reply llvm/include/llvm/MC/CAS/MCCASObjectV1.h]: Ah yes!

File: llvm/include/llvm/MC/CAS/MCCASObjectV1.def (hunk 8 above)
Comment (rastogishubham) [reply llvm/include/llvm/MC/CAS/MCCASObjectV1.def:21]: Done!

File: llvm/lib/MC/MCCASObjectV1.cpp (hunk 9 above)
Comment (rastogishubham) [reply llvm/lib/MC/MCCASObjectV1.cpp]: Oh I am blind, I see what is happening, you are right!

**Comment by rastogishubham** [comment]
@swift-ci please test

**Comment by rastogishubham** [comment]
@swift-ci please test

**Review by felipepiovezan — APPROVED** [review]

File: llvm/lib/MC/MCCASObjectV1.cpp [hunk 14]
```diff
@@ -1885,9 +2045,11 @@ DebugAbbrevOffsetsRefAdaptor::encodeOffsets(ArrayRef<size_t> Offsets) {
 }
 
 static void partitionCUDie(
+    MCCASBuilder &Builder,
```
Comment (felipepiovezan) [inline llvm/lib/MC/MCCASObjectV1.cpp]: is Builder used here?

Comment (rastogishubham) [reply llvm/lib/MC/MCCASObjectV1.cpp]: AHHH! Since I did the refactor with the review comments, it is no longer needed, my bad!

**Comment by rastogishubham** [comment]
@swift-ci please test
